### DISCUSSION TITLE
Local build works, but errors out on vercel - build logs unclear - best way to debug?

### DISCUSSION THREAD

**Discussion by uneatenbreakfast** [vercel/next.js#64188] [Help]
### Summary

I have a next 14.1.0 app that builds without errors locally, but when I deploy onto vercel, it always fails at the build step. It seems building locally is different from building on vercel and I can't seem to track down where the issue is.

On vercel the build log shows this:

```

TypeError: Cannot read properties of undefined (reading 'find')
--
23:48:26.558 | at o (/vercel/path0/.next/server/chunks/539.js:1:32070)
23:48:26.558 | at /vercel/path0/.next/server/chunks/539.js:1:32144
23:48:26.558 | at Array.map (<anonymous>)
23:48:26.558 | at S (/vercel/path0/.next/server/chunks/539.js:1:32137)
23:48:26.558 | at nM (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:47419)
23:48:26.558 | at nN (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:64674)
23:48:26.558 | at nB (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:67657)
23:48:26.558 | at nF (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:66824)
23:48:26.558 | at nN (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:64990)
23:48:26.558 | at nB (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:67657)
23:48:26.558 |  
23:48:26.558 | Error occurred prerendering page "/cart". Read more: https://nextjs.org/docs/messages/prerender-error
23:48:26.559 |  
23:48:26.559 | TypeError: Cannot read properties of undefined (reading 'find')
23:48:26.559 | at o (/vercel/path0/.next/server/chunks/539.js:1:32070)
23:48:26.559 | at /vercel/path0/.next/server/chunks/539.js:1:32144
23:48:26.559 | at Array.map (<anonymous>)
23:48:26.559 | at S (/vercel/path0/.next/server/chunks/539.js:1:32137)
23:48:26.559 | at nM (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:47419)
23:48:26.559 | at nN (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:64674)
23:48:26.559 | at nB (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:67657)
23:48:26.559 | at nF (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:66824)
23:48:26.559 | at nN (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:64990)
23:48:26.559 | at nB (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:67657)

```

So I presume that there is it trying to search an array that is undefined and is failing the "find" operation - but this error isn't clear enough for me to pinpoint exactly where it's having the problem, as it doesn't happen locally.

What would you guys recommend I do here, short of finding every .find/.map operation in my solution and handling for undefined?

**Comment by icyJoseph** [comment]
I remember that there's some process that runs after your build, which is the Vercel API basically preparing to run your code.

I wonder how is `cart` pre-rendering? Or is this happening to all pages? Are you using `NextResponse.error` somewhere?

**Reply by uneatenbreakfast** [reply]
Is there a way to replicate that process locally? I could comment out chunks of my codebase until I can isolate the error locally, but otherwise having to push to vercel to test each change is tiresome.

For this route `cart` - I have 2 client components under the root level react fragment element. Do client components get run during the build process? I thought they didn't, so really don't understand what code is being run during this particular build on vercel.

**Reply by icyJoseph** [reply]
Hi,

Yeah, client components are SSR'd, and pre-rendered during build time. 

- https://github.com/reactwg/server-components/discussions/4

Many tech influencers, and to some degree the naming, has lead to this massive pitfall, where people think of Client components, as rendering client side only, but rather, it is distinction to the bundler, telling it that the code for this components has to be send to the client.

In other words, Client Components, are the components you've used up till now, and the new thing are Server Components.

**Reply by uneatenbreakfast** [reply]
Thanks for that - this helps make this clearer in some aspects. Although it means I'm back to removing/testing every array find() method I come across to see where the issue might be with the vercel build.

Surely I'm not the only one with this issue with an ambiguous build error - what are others doing to debug?

**Reply by icyJoseph** [reply]
But you are 100% sure, that building locally, not dev mode, works?

Most likely you have a list that's undefined. You'd need to put assertions in your code, that's the production way to go about this, but this can also be rooted in other problems, and that's what makes it hard to give pointers.

**Reply by uneatenbreakfast** [reply]
I'm using the default build method `npm run build` that comes with `create next app`.

Perhaps is there some kind of production env variable that is somehow detected when running locally? Or something vercel adds to the build step that somehow makes it different from local?

**Reply by icyJoseph** [reply]
What Node.js version do you have locally, and which one is setup in your Vercel dashboard for this project?

**Reply by uneatenbreakfast** [reply]
It's 20.x in my dashboard and it's v20.10.0 locally.

**Reply by JulianKingman** [reply]
Did you ever find a way to debug this? In the same boat here :\

**Reply by kunlunli** [reply]
Hey guys, consider integrating Sentry! It's a very convenient tool to help you identify both build time and runtime errors. Here is an example:

This is the build log I have in my local console, very vague and impossible to trace, like your case:
`TypeError: Cannot read properties of null (reading 'default')
    at s (Project/github/orgnization-repo/Web-Next/commently-web/.next/server/chunks/937.js:1:383928)
    at (Project/github/orgnization-repo/Web-Next/commently-web/.next/server/chunks/937.js:1:386132
    at l (Project/github/orgnization-repo/Web-Next/commently-web/.next/server/chunks/937.js:1:386177)
    at /Project/github/orgnization-repo/Web-Next/commently-web/.next/server/chunks/937.js:1:361480
    at nj (Project/github/orgnization-repo/Web-Next/commently-web/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:46251)
    at nM (Project/github/orgnization-repo/Web-Next/commently-web/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:60839)
    at nN (Project/github/orgnization-repo/Web-Next/commently-web/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:64546)
    at nB (Project/github/orgnization-repo/Web-Next/commently-web/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:67538)
    at nD (Project/github/orgnization-repo/Web-Next/commently-web/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:66680)
    at nN (Project/github/orgnization-repo/Web-Next/commently-web/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:64853)`

Now, after the config, it will show up in Sentry dashboard like this:
`../../src/shared/lib/get-img-props.ts in isStaticRequire at line 101:33
In App

../../src/shared/lib/get-img-props.ts in getImgProps at line 113:6
In App

../../src/client/image-component.tsx in Image at line 397:20`

You can easily figure out the problem and locate the issue.

**Reply by JulianKingman** [reply]
For some reason Sentry doesn't give me build logs. Working on that with the Sentry team.

Is this something the next/vercel team is working on? Out of all the things that make a framework easy to deal with, exposed and understandable errors would be right up there at the top of my list. I feel like I'm missing something, it shouldn't be this hard to find my errors 🤔

**Reply by JulianKingman** [reply]
@kunlunli chatted with Sentry support, apparently Sentry doesn't provide build error tracking. Was there something else you did that might have enabled the source maps?
